src/main/java/net/sf/jabref/gui/actions/CleanupAction.java

```diff
@@ -0,0 +1,174 @@
+/*  Copyright (C) 2012-2015 JabRef contributors.
+    This program is free software; you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation; either version 2 of the License, or
+    (at your option) any later version.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License along
+    with this program; if not, write to the Free Software Foundation, Inc.,
+    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+package net.sf.jabref.gui.actions;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+
+import javax.swing.JOptionPane;
+import net.sf.jabref.*;
+import net.sf.jabref.gui.*;
+import net.sf.jabref.gui.worker.AbstractWorker;
+import net.sf.jabref.gui.undo.NamedCompound;
+import net.sf.jabref.gui.undo.UndoableFieldChange;
+
+import net.sf.jabref.logic.FieldChange;
+import net.sf.jabref.logic.cleanup.CleanupPreset;
+import net.sf.jabref.logic.cleanup.CleanupWorker;
+import net.sf.jabref.logic.l10n.Localization;
+import net.sf.jabref.logic.preferences.PreferenceStorage;
+import net.sf.jabref.model.entry.BibEntry;
+
+public class CleanupAction extends AbstractWorker {
+
+    private final BasePanel panel;
+    private final JabRefFrame frame;
+
+    /**
+     * Global variable to count unsuccessful renames
+     */
+    private int unsuccessfulRenames;
+
+    private boolean cancelled;
+    private int modifiedEntriesCount;
+    private final PreferenceStorage preferences;
+    private final CleanupPresetPanel presetPanel;
+
+
+    public CleanupAction(BasePanel panel, PreferenceStorage preferences) {
+        this.panel = panel;
+        this.frame = panel.frame();
+        this.preferences = Objects.requireNonNull(preferences);
+        this.presetPanel = new CleanupPresetPanel(preferences.retrieveCleanupPreset());
+    }
+
+    @Override
+    public void init() {
+        cancelled = false;
+        modifiedEntriesCount = 0;
+        int numSelected = panel.getSelectedEntries().length;
+        if (numSelected == 0) { // None selected. Inform the user to select entries first.
+            JOptionPane.showMessageDialog(frame, Localization.lang("First select entries to clean up."),
+                    Localization.lang("Cleanup entry"), JOptionPane.INFORMATION_MESSAGE);
+            cancelled = true;
+            return;
+        }
+        frame.block();
+        panel.output(Localization.lang("Doing a cleanup for %0 entries...", Integer.toString(numSelected)));
+    }
+
+    @Override
+    public void run() {
+        if (cancelled) {
```

Comment:
I would move some of the code in the init part and some of it in the update part. You can have a look at the `SearchWorker`, which separate this correctly. The issue is that the run method is always executed outside of the EDT, and the update method always on the EDT. And swing only works correctly if all methods on all swing classes are executed on the EDT. 

This change may require a lot of work. If you think this is too much, we can leave it as is and improve that in another PR.